---
title: "Analyze data using ArchR"
author: "Mingbo Cheng"
date: '`r format(Sys.Date(), "%Y-%B-%d")`'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(warning = F, message = F, echo = T,cache=F)
library(ggplot2)
library(stringr)
library(cowplot)
library(magrittr)
library(WriteXLS)
library(tidyr)
library(dplyr)
library(plotly)
library(cluster)
library(cowplot)
library(gridExtra)
library(viridis)
library(GenomicRanges)
library(GenomeInfoDb)
library(data.table)
library(ArchR)
library(glue)
```

```{r set_parameters, echo=FALSE}
## set parameters
set.seed(42)
addArchRThreads(threads =100)
addArchRGenome("mm10")


cols.celltypes <- c(
    'Fibroblasts' = '#f032e6',
    'Myofibroblasts' = '#911eb4',
    'Pericytes/vSMC' = '#e6beff')

Add_A_Matrix <- function(
    project = proj,
    mtx = mtx,
    vname = "z",
    name = "AMatrix"
){
  ArrowFiles <- getArrowFiles(proj)
  nms <- names(ArrowFiles)
  for(nm in nms){
    a_cellnm <- colnames(mtx)[grepl(nm, colnames(mtx))]
    message(nm, " ", length(a_cellnm))
    a_mtx <- mtx[, a_cellnm]
    ArrowFile <- ArrowFiles[nm]
    featureDF <-data.frame(seqnames = vname,
                           idx = seq_len(nrow(a_mtx)),
                           name = rownames(a_mtx),
                           stringsAsFactors = FALSE)

    #Initialize
   o <- ArchR:::.initializeMat(
    ArrowFile = ArrowFile,
    Group = name,
    Class = "double",
    Units =  vname,
    cellNames = a_cellnm,
    params = vname,
    featureDF = featureDF,
    force = TRUE)

   o <- ArchR:::.addMatToArrow(
        mat = as(a_mtx,"dgCMatrix"),
        ArrowFile = ArrowFile,
        Group = paste0(name, "/", vname),
        binarize = FALSE,
        addColSums = TRUE,
        addRowSums = TRUE,
        addRowVarsLog2 = TRUE)
    o <- h5write(obj = "Finished", file = ArrowFile, name = paste0(name,"/Info/Completed"))
  }
  return(proj)
}

```

### Reading data
```{r}
proj <- loadArchRProject(path = "./Fib", showLogo = FALSE)
proj <- addImputeWeights(proj)
```



```{r}
trajGSM1 <- getTrajectory(ArchRProj = proj, 
                         name = "palantir_Fibroblasts", 
                         useMatrix = "GeneScoreMatrix", 
                         log2Norm = TRUE)

trajMM1 <- getTrajectory(ArchRProj = proj, 
                         name = "palantir_Fibroblasts", 
                         useMatrix = "MotifMatrix", 
                         log2Norm = FALSE,
                         scaleTo = NULL)



trajGSM2 <- getTrajectory(ArchRProj = proj, 
                         name = "palantir_Pericytes_vSMC", 
                         useMatrix = "GeneScoreMatrix", 
                         log2Norm = TRUE)


trajMM2 <- getTrajectory(ArchRProj = proj, 
                         name = "palantir_Pericytes_vSMC", 
                         useMatrix = "MotifMatrix", 
                         log2Norm = FALSE,
                         scaleTo = NULL)


motif_mtx <- getMatrixFromProject(proj, "MotifMatrix")
mtx <- assays(motif_mtx)[["z"]]
proj <- Add_A_Matrix(proj, mtx, name="MotifZMatrix")



corGSM_MM1 <- correlateTrajectories(trajGSM1, 
                                    trajMM1,
                                    corCutOff = 0.0,
                                    varCutOff1 = 0.8,
                                    varCutOff2 = 0.8)

corGSM_MM1[[1]]
corGSM_MM1[[1]] <- subset(corGSM_MM1[[1]], FDR < 0.05)

trajGSM1 <- trajGSM1[corGSM_MM1[[1]]$name1, ]
trajMM1 <- trajMM1[corGSM_MM1[[1]]$name2, ]

trajCombined <- trajGSM1
assay(trajCombined, withDimnames=FALSE) <- t(apply(assay(trajGSM1), 1, scale)) + t(apply(assay(trajMM1), 1, scale))

combinedMat <- plotTrajectoryHeatmap(trajCombined,
                                     returnMat = TRUE,
                                     varCutOff = 0)

rowOrder <- match(rownames(combinedMat), rownames(trajGSM1))
rowOrder <- unique(rowOrder)



gene_pos <- rownames(trajGSM1)
genes <- unique(elementMetadata(trajGSM1[rowOrder])$name)
#genes <- unique(sapply(stringr::str_split(corGSM_MM1[[1]]$name1, ":"), function(x) x[2])) 

###-------------strange, will failed if ran with the next plot loop, but succeeded if stop here ------
pdf("../figure/fib/corr_fib_genes_palantir.pdf", width=10, height=4)
for(g in genes){
	p1 <- plotTrajectory(proj, trajectory = "palantir_Fibroblasts", 
				               colorBy = "GeneScoreMatrix",
                               name = g, 
                               quantCut = c(0.01, 0.95),
                               embedding = "palantir",
                               size = 0.5,
                               title="fibroblasts",
				               continuousSet = "blueYellow")


	print(plot_grid(p1[[1]], p1[[2]]), ncol=2)
}
dev.off()


pdf(file.path("../figure/fib", glue("corr_motif_fib_trajectory.pdf")), width=10, height=4)
motifs_pos <- rownames(trajMM1)
motifs <- unique(elementMetadata(trajMM1[rowOrder])$name)
for (mtf in motifs) {
    mtf_pos <- rownames(trajMM1)[which(grepl(mtf, rownames(trajMM1)))[1]]
    p1 <- plotTrajectory(proj,
                    trajectory = "palantir_Fibroblasts",
                    colorBy = "MotifZMatrix",
                    name = mtf_pos,
                    continuousSet = "blueYellow",
                    embedding = "palantir",
                    title = "palantir_Pericytes_vSMC")

    ggAlignPlots(p1[[1]],
                 p1[[2]],
                type="h")
}
dev.off()


```


```{r}
corGSM_MM2 <- correlateTrajectories(trajGSM2, 
                                    trajMM2,
                                    corCutOff = 0.0)

corGSM_MM2[[1]]
corGSM_MM2[[1]] <- subset(corGSM_MM2[[1]], FDR < 0.05)

trajGSM2 <- trajGSM2[corGSM_MM2[[1]]$name1, ]
trajMM2 <- trajMM2[corGSM_MM2[[1]]$name2, ]
trajCombined <- trajGSM2
assay(trajCombined, withDimnames=FALSE) <- t(apply(assay(trajGSM2), 1, scale)) + t(apply(assay(trajMM2), 1, scale))

combinedMat <- plotTrajectoryHeatmap(trajCombined,
                                     returnMat = TRUE,
                                     varCutOff = 0)

rowOrder <- match(rownames(combinedMat), rownames(trajGSM2))
rowOrder <- unique(rowOrder)


#genes <- unique(sapply(stringr::str_split(corGSM_MM2[[1]]$name1, ":"), function(x) x[2])) 
gene_pos <- rownames(trajGSM2)
genes <- unique(elementMetadata(trajGSM2[rowOrder])$name)

#genes <- c(genes, c("Smad3", "Bach2", "Nfix", "Nfkb1", "Atoh8"))
###-------------strange, will failed if ran with the next plot loop, but succeeded if stop here ------
pdf("../figure/fib/corr_pericytes_genes_palantir.pdf", width=10, height=4)
i=1
for(g in genes){
	message(g, " ", i)
    p1 <- plotTrajectory(proj, trajectory = "palantir_Pericytes_vSMC", 
				               colorBy = "GeneScoreMatrix",
                               name = g, 
                               quantCut = c(0.01, 0.95),
                               embedding = "palantir",
                               size = 0.5,
                               title="pericytes",
				               continuousSet = "blueYellow")


	print(plot_grid(p1[[1]], p1[[2]]), ncol=2)
 	i = i+1
}
dev.off()

motifs_pos <- rownames(trajMM2)
motifs <- unique(elementMetadata(trajMM2[rowOrder])$name)

pdf(file.path("../figure/fib", glue("corr_motif_peri_trajectory.pdf")), width=10, height=4 )
for (mtf in motifs) {
    mtf_pos <- rownames(trajMM2)[which(grepl(mtf, rownames(trajMM2)))[1]]
    p1 <- plotTrajectory(proj,
                    trajectory = "palantir_Pericytes_vSMC",
                    colorBy = "MotifZMatrix",
                    name = mtf_pos,
                    continuousSet = "blueYellow",
                    embedding = "palantir",
                    title = "palantir_Pericytes_vSMC")

    ggAlignPlots(p1[[1]],
                 p1[[2]],
                type="h")
}
dev.off()




#pdf("genes__palantir.pdf", width=10, height=4)
#for(g in genes){
#   p <- plotEmbedding(
#    	ArchRProj = proj, 
#    	colorBy = "GeneScoreMatrix", 
#    	name = g, 
#    	embedding = "palantir",
#    	size = 1,
#    	quantCut = c(0.01, 0.95))
#}
#dev.off()


#
#                    trajectory = "palantir_Pericytes_vSMC", 
#                    colorBy = "GeneScoreMatrix", 
#                    name = gene, 
#                    continuousSet = "blueYellow",
#                    embedding = "palantir",
#                    title = "palantir_Pericytes_vSMC")
#
#        p1 <- plotTrajectory(proj,
#                    trajectory = "palantir_Fibroblasts",
#                    colorBy = "cellColData",
#                    name = "palantir_Fibroblasts",
#                    embedding = "palantir",
#                    size = 0.5,
#                    quantCut = c(0.01, 0.95),
#                    title = "Fibroblasts")
#
```


